src/db.c

```diff
@@ -46,6 +48,80 @@ void updateLFU(robj *val) {
     val->lru = (LFUGetTimeInMinutes()<<8) | counter;
 }
 
+/* 
+ * Update histogram of keys-sizes
+ * 
+ * It is used to track the distribution of key sizes in the dataset. It is updated 
+ * every time key's length is modified. Available to user via INFO command. 
+ * 
+ * The histogram is a base-2 logarithmic histogram, with 64 bins. The i'th bin 
+ * represents the number of keys with a size in the range 2^i and 2^(i+1) 
+ * exclusive. oldLen/newLen must be smaller than 2^63. Each data type has its
+ * own histogram and it is per database (In addition, there is histogram per 
+ * slot for future use).
+ * 
+ * Examples to LEN values and corresponding bins in histogram: 
+ *               [1,2)->0 [2,4)->1 [4,8)->2 [8,16)->3
+ */
+void updateKeysizesHist(redisDb *db, int didx, uint32_t type, uint64_t oldLen, uint64_t newLen) {
+    uint64_t dummyHist[MAX_KEYSIZES_BINS];
+
+    if  (unlikely(type >= OBJ_TYPE_BASIC_MAX)) 
+        return;
+
+    kvstoreDictMetadata *metadata = kvstoreGetDictMetadata(db->keys, didx);
+
+    /* If following key deletion, If it is last one in slot's dict, then 
+     * slot's dict might get released as well. Verify if metadata is available. */
+    uint64_t *dictHist = (metadata != NULL) ? metadata->keysizes_hist[type] : dummyHist;
+
+    uint64_t *kvstoreHist = kvstoreGetMetadata(db->keys)->keysizes_hist[type];
+
+    if (oldLen != 0) {
+        int oLdBin = log2ceil(oldLen);
+        debugServerAssertWithInfo(server.current_client, NULL, oLdBin < MAX_KEYSIZES_BINS);
+        --dictHist[oLdBin];
+        --kvstoreHist[oLdBin];
+    }
+    
+    if (newLen != 0) {
+        int newBin = log2ceil(newLen);
+        debugServerAssertWithInfo(server.current_client, NULL, newBin < MAX_KEYSIZES_BINS);
+        ++dictHist[newBin];
+        ++kvstoreHist[newBin];
+    }
+}
+
+/* Remove key statistic from keysizes histogram */
+void removeKeysizesHist(redisDb *db, int didx, robj *val) {
+    uint64_t len;
+    switch (val->type)
```

Comment:
It is too short and too specific to be bothered about. Please note that stream is not included in the party.

Reply:
now we have three same code, so i also prefer to add a new function to get the length of key.
although stream is not implemented, but it will be skipped in updateKeysizesHist(), which i think it's fine.

Reply:
The code looks better than i expected. Now removeKeysizesHist() and addKeysizesHist() become a trivial single line call to updateKeysizesHist + length. So we can discard those functions as well!